app/cdap/components/NamespaceAdmin/store/ActionCreator.ts

```diff
@@ -189,6 +196,132 @@ export function deleteConnection(conn: IConnection) {
   });
 }
 
+export const getSourceControlManagement = (namespace) => {
+  const params = { namespace };
+  MyNamespaceApi.getSourceControlManagement(params)
+    .pipe(
+      catchError((err) => {
+        Store.dispatch({
+          type: NamespaceAdminActions.setSourceControlManagementConfig,
+          payload: {
+            sourceControlManagementConfig: null,
+          },
+        });
+        throw err;
+      })
+    )
+    .subscribe((res) => {
+      // after getting the saved config, we need to fetch the token using the tokenName
+      const config = res.config;
+      MySecureKeyApi.getSecureData({ ...params, key: config.auth.tokenName }).subscribe((token) => {
+        config.auth.token = token;
+        Store.dispatch({
+          type: NamespaceAdminActions.setSourceControlManagementConfig,
+          payload: {
+            sourceControlManagementConfig: config,
+          },
+        });
+      });
+    });
+};
+
+export const addSourceControlManagement = (
+  namespace,
+  formState: ISourceControlManagementConfig,
+  callbackFn: any,
+  validate: boolean = false
+) => {
+  const params = { namespace };
+  if (validate) {
+    // validate the connection
+    // TODO: currently it requires to save the token to secure store first.
+    // In the future we might want to test connection on the fly
+    MySecureKeyApi.put(
+      { ...params, key: formState.auth.tokenName },
+      {
+        data: formState.auth.token,
+      }
+    )
+      .pipe(
+        catchError((err) => {
+          callbackFn({
+            type: 'SET_ERROR',
+            payload: {
+              error: err.message,
+            },
+          });
+          throw err;
+        })
+      )
+      .subscribe(() => {
+        MyNamespaceApi.setSourceControlManagement(
+          params,
+          getBodyForSubmit(formState, validate)
+        ).subscribe(
+          () => {
+            // success
+            callbackFn({
+              type: 'SET_SUCCESS',
+            });
+          },
+          (err) => {
+            callbackFn({
+              type: 'SET_ERROR',
+              payload: {
+                error: err.message,
+              },
+            });
+          }
+        );
+      });
+    return;
+  }
+  Observable.forkJoin(
+    MyNamespaceApi.setSourceControlManagement(params, getBodyForSubmit(formState)),
+    MySecureKeyApi.put(
+      { ...params, key: formState.auth.tokenName },
+      {
+        data: formState.auth.token,
+      }
+    )
+  ).subscribe(() => {
+    Store.dispatch({
+      type: NamespaceAdminActions.setSourceControlManagementConfig,
+      payload: {
+        sourceControlManagementConfig: formState,
+      },
+    });
+    callbackFn();
+  });
+};
+
+export const deleteSourceControlManagement = (
+  namespace,
+  formState: ISourceControlManagementConfig,
+  callbackFn: any
+) => {
+  const params = { namespace };
+  Observable.forkJoin(
```

Comment:
These Observables will run in parallel. Is there any issue if the auth token is deleted before the SCM config? Or if one errors but the other doesn't?

Reply:
good point. probably should delete scm config first then delete auth token